// babel-plugin-styled-components-px2rem/lib/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ConfigurationManager = /** @class */ (function () {
    function ConfigurationManager() {
        this._config = ConfigurationManager.defaultConfiguration;
    }
    Object.defineProperty(ConfigurationManager.prototype, "config", {
        get: function () {
            return this._config;
        },
        enumerable: true,
        configurable: true
    });
    ConfigurationManager.prototype.updateConfig = function (config) {
        if (config) {
            this._config = Object.assign({}, ConfigurationManager.defaultConfiguration, config);
        }
    };
    ConfigurationManager.defaultConfiguration = {
        rootValue: 100,
        unitPrecision: 5,
        minPixelValue: 0,
        multiplier: 1,
        tags: ['styled', 'css', 'createGlobalStyle', 'keyframes'],
        propWhiteList: [],
        propBlackList: [],
        exclude: false,
        selectorBlackList: [],
        ignoreIdentifier: false,
        replace: true,
        mediaQuery: false,
        transformRuntime: false,
    };
    return ConfigurationManager;
}());
exports.ConfigurationManager = ConfigurationManager;
exports.default = new ConfigurationManager();
